velox/functions/prestosql/window/NthValue.cpp

```diff
@@ -0,0 +1,189 @@
+    partitionOffset_ = 0;
+  }
+
+  template <bool constantOffset>
+  void getRowNumbersForOffset(
+      vector_size_t numRows,
+      const vector_size_t* frameStarts,
+      const vector_size_t* frameEnds) {
+    // The rowNumbers map for each output row, as per nth_value function
+    // semantics, the rowNumber (relative to the start of the partition) from
+    // which the input value should be copied.
+    // A rowNumber of -1 is for nullptr in the result.
+    rowNumbersVector_.resize(numRows);
+
+    if constexpr (constantOffset) {
+      if (isConstantOffsetNull_) {
+        for (int i = 0; i < numRows; i++) {
+          rowNumbersVector_[i] = -1;
+        }
+        return;
+      }
+      for (int i = 0; i < numRows; i++) {
+        auto frameStart = frameStarts[i];
+        auto frameEnd = frameEnds[i];
+        auto rowNumber = frameStart + constantOffset_ - 1;
+        rowNumbersVector_[i] = rowNumber <= frameEnd ? rowNumber : -1;
+      }
+      return;
+    }
+
+    offsetsVector_->resize(numRows);
+    partition_->extractColumn(
+        args_[1].argIndex, partitionOffset_, numRows, 0, offsetsVector_);
+    for (int i = 0; i < numRows; i++) {
+      if (offsetsFlatVector_->isNullAt(i)) {
+        rowNumbersVector_[i] = -1;
+      } else {
+        auto offset = offsetsFlatVector_->valueAt(i);
+        VELOX_USER_CHECK_GE(offset, 1, "Offset must be at least 1");
+        auto frameStart = frameStarts[i];
+        auto frameEnd = frameEnds[i];
+
+        auto rowNumber = frameStart + offset - 1;
+        rowNumbersVector_[i] = rowNumber <= frameEnd ? rowNumber : -1;
+      }
+    }
+  }
+
+  void apply(
+      const BufferPtr& /*peerGroupStarts*/,
+      const BufferPtr& /*peerGroupEnds*/,
+      const BufferPtr& frameStarts,
+      const BufferPtr& frameEnds,
+      int32_t resultOffset,
+      const VectorPtr& result) {
+    auto numRows = frameStarts->size() / sizeof(vector_size_t);
+    auto frameStartsPtr = frameStarts->as<vector_size_t>();
+    auto frameEndsPtr = frameEnds->as<vector_size_t>();
+
+    if (isConstantOffset_) {
+      getRowNumbersForOffset<true>(numRows, frameStartsPtr, frameEndsPtr);
+    } else {
+      getRowNumbersForOffset<false>(numRows, frameStartsPtr, frameEndsPtr);
+    }
+
+    auto rowNumbersRange = folly::Range(rowNumbersVector_.data(), numRows);
+    partition_->extractColumn(
+        args_[0].argIndex, rowNumbersRange, resultOffset, result);
+
+    partitionOffset_ += numRows;
+  }
+
+ private:
+  // This needs to be a copy of the function args as they could go out of
+  // scope.
+  const std::vector<exec::WindowFunctionArg> args_;
+
+  const exec::WindowPartition* partition_;
+
+  // These fields are set if the offset argument is a constant value.
+  bool isConstantOffset_ = false;
+  bool isConstantOffsetNull_ = false;
+  int64_t constantOffset_;
+
+  // This vector is used to extract values of the offset argument column
+  // (if not a constant offset value).
+  VectorPtr offsetsVector_;
```

Comment:
I wonder if it makes sense to drop Vector suffix for readability, i.e. offsets_, flatOffsets_, rowsNumbers. Using Vector is a bit confusing because sometimes it it velox::Vector and sometimes it is std::vector.